enterprise/utils.py

```diff
@@ -2419,7 +2420,4 @@ def ensure_course_enrollment_is_allowed(course_id, email, enrollment_api_client)
 
     course_details = enrollment_api_client.get_course_details(course_id)
     if course_details["invite_only"]:
-        CourseEnrollmentAllowed.objects.update_or_create(
```

Comment:
I [tried to re-use this helper as is](https://github.com/open-craft/edx-enterprise/commit/6d0f75ef760162ffa7799e60930db28c14d3f117), but it was locking a DB row causing enrollment to fail [here](https://github.com/openedx/edx-platform/blob/5f9a30e0f24b7d136e63c10481654ec5a85b9d82/common/djangoapps/student/models/course_enrollment.py#L359-L364), because there we're trying to modify the locked row.

Reply:
@0x29a It's surprising that we haven't run into this race condition before. Any idea why it might be occuring now?

Irrespective of the reasons, I think using the API is a fine approach to take, given that we get the course details in the previous statement using the API and not an import. It looks like the `enrollment_allowed` is a recent API addition (added 6/7 months ago). We would have probably used it if it existed earlier, as this allows creating CEA objects from other services like an admin MFE as well.

Reply:
Thanks for checking this, @tecoholic.

>Any idea why it might be occuring now?

This was bothering me as well, but I was unable to figure this out back then. It turns out that the view modified in https://github.com/openedx/edx-enterprise/pull/1813 is [not atomic](https://github.com/openedx/edx-enterprise/blob/e6aab238412c6ecdd3e562acb96922d38757d503/enterprise/views.py#L1256). And the view modified in https://github.com/open-craft/edx-enterprise/pull/14 doesn't send any internal HTTP requests in fact, so CEA objects are manipulated within a single transaction. We can also change `View` with `NonAtomicView` [here](https://github.com/open-craft/edx-enterprise/blob/5927941072bd397fd7c4a4143ea2d45235d193f4/enterprise/views.py#L324), and this will simplify this fix, but I'm not sure how risky this is.